tokio-net/src/driver/reactor/sharded_slab/slab.rs

```diff
@@ -0,0 +1,237 @@
+use super::*;
+use std::fmt;
+
+use crate::sync::atomic::Ordering;
+
+/// A sharded slab.
+pub(crate) struct Slab {
+    shards: Box<[Shard]>,
+}
+
+// ┌─────────────┐      ┌────────┐
+// │ page 1      │      │        │
+// ├─────────────┤ ┌───▶│  next──┼─┐
+// │ page 2      │ │    ├────────┤ │
+// │             │ │    │XXXXXXXX│ │
+// │ local_free──┼─┘    ├────────┤ │
+// │ global_free─┼─┐    │        │◀┘
+// ├─────────────┤ └───▶│  next──┼─┐
+// │   page 3    │      ├────────┤ │
+// └─────────────┘      │XXXXXXXX│ │
+//       ...            ├────────┤ │
+// ┌─────────────┐      │XXXXXXXX│ │
+// │ page n      │      ├────────┤ │
+// └─────────────┘      │        │◀┘
+//                      │  next──┼───▶
+//                      ├────────┤
+//                      │XXXXXXXX│
+//                      └────────┘
+//                         ...
+pub(super) struct Shard {
+    #[cfg(debug_assertions)]
+    tid: usize,
+    /// The local free list for each page.
+    ///
+    /// These are only ever accessed from this shard's thread, so they are
+    /// stored separately from the shared state for the page that can be
+    /// accessed concurrently, to minimize false sharing.
+    local: Box<[page::Local]>,
+    /// The shared state for each page in this shard.
+    ///
+    /// This consists of the page's metadata (size, previous size), remote free
+    /// list, and a pointer to the actual array backing that page.
+    shared: Box<[page::Shared]>,
+}
+
+pub(crate) const TOKEN_SHIFT: usize = Tid::SHIFT + Tid::LEN;
+pub(crate) const MAX_SOURCES: usize = (1 << TOKEN_SHIFT) - 1;
+
+impl Slab {
+    /// Returns a new slab with the default configuration parameters.
+    pub(crate) fn new() -> Self {
```

Comment:
The # of max threads should be passed in here instead of using the const. The const should be for the max MAX threads :).